drivers/dma/dma_max32.c

```diff
@@ -0,0 +1,341 @@
+{
+	switch (addr_adj) {
+	case DMA_ADDR_ADJ_NO_CHANGE:
+		return 0;
+	case DMA_ADDR_ADJ_INCREMENT:
+		return 1;
+	default:
+		LOG_ERR("Invalid DMA address adjust - must be NO_CHANGE (0) or INCREMENT (1)");
+		return 0;
+	}
+}
+
+static inline int get_dma_ch_index(mxc_dma_regs_t *dma,  uint8_t ch)
+{
+	return (ch + MXC_DMA_GET_IDX(dma) * (MXC_DMA_CHANNELS / MXC_DMA_INSTANCES));
+}
+
+static int api_config(const struct device *dev, uint32_t channel,
+				   struct dma_config *config)
+{
+	int ret = 0;
+	const struct max32_dma_config *cfg = dev->config;
+	struct max32_dma_data *data = dev->data;
+	uint32_t ch;
+
+	if (channel >= cfg->channels) {
+		LOG_ERR("Invalid DMA channel - must be < %" PRIu32 " (%" PRIu32 ")", cfg->channels,
+			channel);
+		return -EINVAL;
+	}
+
+	ch = get_dma_ch_index(cfg->regs, channel);
+
+	/* DMA Channel Config */
+	mxc_dma_config_t mxc_dma_cfg;
+
+	mxc_dma_cfg.ch = ch;
+	mxc_dma_cfg.reqsel = config->dma_slot << ADI_MAX32_DMA_CFG_REQ_POS;
+	if (((get_dma_width(config->source_data_size)) < 0) ||
+	      ((get_dma_width(config->dest_data_size)) < 0)) {
+		return -EINVAL;
+	}
+	mxc_dma_cfg.srcwd = get_dma_width(config->source_data_size);
+	mxc_dma_cfg.dstwd = get_dma_width(config->dest_data_size);
+	mxc_dma_cfg.srcinc_en = get_dma_addr_adj(config->head_block->source_addr_adj);
+	mxc_dma_cfg.dstinc_en = get_dma_addr_adj(config->head_block->dest_addr_adj);
+
+	/* DMA Channel Advanced Config */
+	mxc_dma_adv_config_t mxc_dma_cfg_adv;
+
+	mxc_dma_cfg_adv.ch = ch;
+	if (!is_valid_dma_ch_prio(config->channel_priority)) {
+		return -EINVAL;
+	}
+	mxc_dma_cfg_adv.prio = config->channel_priority;
+	mxc_dma_cfg_adv.reqwait_en = 0;
+	mxc_dma_cfg_adv.tosel = MXC_DMA_TIMEOUT_4_CLK;
+	mxc_dma_cfg_adv.pssel = MXC_DMA_PRESCALE_DISABLE;
+	mxc_dma_cfg_adv.burst_size = config->source_burst_length;
+
+	/* DMA Transfer Config */
+	mxc_dma_srcdst_t txfer;
+
+	txfer.ch = ch;
+	txfer.source = (void *)config->head_block->source_address;
+	txfer.dest = (void *)config->head_block->dest_address;
+	txfer.len = config->head_block->block_size;
+
+	ret = MXC_DMA_ConfigChannel(mxc_dma_cfg, txfer);
+	if (ret != E_NO_ERROR) {
+		return ret;
+	}
+
+	ret = MXC_DMA_AdvConfigChannel(mxc_dma_cfg_adv);
+	if (ret) {
+		return ret;
+	}
+
+	/* Enable interrupts for the DMA peripheral */
+	ret = MXC_DMA_EnableInt(ch);
+	if (ret != E_NO_ERROR) {
+		return ret;
+	}
+
+	/* Enable complete and count-to-zero interrupts for the channel */
+	ret = MXC_DMA_ChannelEnableInt(ch,
+				       ADI_MAX32_DMA_CTRL_DIS_IE | ADI_MAX32_DMA_CTRL_CTZIEN);
+	if (ret != E_NO_ERROR) {
+		return ret;
+	}
+
+	data[channel].callback = config->dma_callback;
+	data[channel].cb_data = config->user_data;
+	data[channel].err_cb_dis = config->error_callback_dis;
+
+	return ret;
+}
+
+static int api_reload(const struct device *dev, uint32_t channel, uint32_t src,
+				   uint32_t dst, size_t size)
+{
+	const struct max32_dma_config *cfg = dev->config;
+	mxc_dma_srcdst_t reload;
+
```

Comment:
What if the channel is currently mid-transfer or previously unconfigured? I see no validation of state here